dawn/prototype/mylibIconLaplace.cpp

```diff
@@ -0,0 +1,597 @@
+
+  // init geometric info for cells
+  for(const auto& c : mesh.faces()) {
+    cell_area(c, level) = CellArea(c);
+  }
+  // init geometric info for vertices
+  for(const auto& v : mesh.vertices()) {
+    dual_cell_area(v, level) = DualCellArea(v);
+  }
+
+  if(dbg_out) {
+    dumpField("laplICONmylib_areaCell.txt", mesh, cell_area, level);
+    dumpField("laplICONmylib_areaCellDual.txt", mesh, dual_cell_area, level);
+  }
+
+  // init edge orientations for vertices and cells
+  auto dot = [](const mylib::Vertex& v1, const mylib::Vertex& v2) {
+    return v1.x() * v2.x() + v1.y() * v2.y();
+  };
+
+  // +1 when the vector from this to the neigh-
+  // bor vertex has the same orientation as the
+  // tangent unit vector of the connecting edge.
+  // -1 otherwise
+  for(const auto& v : mesh.vertices()) {
+    int m_sparse = 0;
+    for(const auto& e : v.edges()) {
+      mylib::Vertex testVec =
+          mylib::Vertex(v.vertex(m_sparse).x() - v.x(), v.vertex(m_sparse).y() - v.y(), -1);
+      mylib::Vertex dual = mylib::Vertex(dual_normal_x(*e, level), dual_normal_y(*e, level), -1);
+      edge_orientation_vertex(v, m_sparse, level) =
+          sgn(dot(mylib::Vertex(v.vertex(m_sparse).x() - v.x(), v.vertex(m_sparse).y() - v.y(), -1),
+                  mylib::Vertex(dual_normal_x(*e, level), dual_normal_y(*e, level), -1)));
+      m_sparse++;
+    }
+  }
+
+  // The orientation of the edge normal vector
+  // (the variable primal normal in the edges ta-
+  // ble) for the cell according to Gauss formula.
+  // It is equal to +1 if the normal to the edge
+  // is outwards from the cell, otherwise is -1.
+  for(const auto& c : mesh.faces()) {
+    int m_sparse = 0;
+    auto [xm, ym] = CellCircumcenter(c);
+    for(const auto& e : c.edges()) {
+      mylib::Vertex vOutside(e->vertex(0).x() - xm, e->vertex(0).y() - ym, -1);
+      edge_orientation_cell(c, m_sparse, level) =
+          sgn(dot(mylib::Vertex(e->vertex(0).x() - xm, e->vertex(0).y() - ym, -1),
+                  mylib::Vertex(primal_normal_x(*e, level), primal_normal_y(*e, level), -1)));
+      m_sparse++;
+      // explanation: the vector cellModpoint -> e.vertex(0) is guaranteed to point outside. The dot
+      // product checks if the edge normal has the same orientation. e.vertex(0) is arbitrary,
+      // vertex(1), or any point on e would work just as well
+    }
+  }
+
+  // init sparse quantities for div and rot
+  for(const auto& v : mesh.vertices()) {
+    int m_sparse = 0;
+    for(const auto& e : v.edges()) {
+      geofac_rot(v, m_sparse, level) = dual_edge_length(*e, level) *
+                                       edge_orientation_vertex(v, m_sparse, level) /
+                                       dual_cell_area(v, level);
+
+      // ptr_int%geofac_rot(jv,je,jb) =                &
+      //    & ptr_patch%edges%dual_edge_length(ile,ibe) * &
+      //    & ptr_patch%verts%edge_orientation(jv,jb,je)/ &
+      //    & ptr_patch%verts%dual_area(jv,jb) * REAL(ifac,wp)
```

Comment:
what is `REAL(ifac,wp)`? should we care?

Reply:
```
    ! Vorticity should have the right sign
    SELECT CASE (ptr_patch%geometry_info%cell_type)
    CASE (3)
      ifac = 1
    CASE (6)
      ifac = -1
    END SELECT
```
-1 for hexagons, +1 for triangles, we are save ;)